src/logging.c

```diff
@@ -761,6 +768,51 @@ int rcutils_logging_get_logger_level(const char * name)
   return rcutils_logging_get_logger_leveln(name, strlen(name));
 }
 
+static rcutils_ret_t add_key_to_hash_map(const char * name, int level, bool set_by_user)
+{
+  // Copy the name that we will store, as there is no guarantee that the caller will keep it around.
+
+  char * copy_name = rcutils_strdup(name, g_rcutils_logging_allocator);
+  if (copy_name == NULL) {
+    // Don't report an error to the error handling machinery; some uses of this function are for
+    // caching so this is not necessarily fatal.
+    return RCUTILS_RET_ERROR;
+  }
+
+  if (set_by_user) {
+    // If the level was set by the user (rather than optimistically cached by the implementation),
+    // mark it here.  When we purge the cache, we'll make sure not to purge these.  The mark we
+    // use is setting the bottom bit; since our levels are 0, 10, 20, 30, 40, and 50, this works.
+    level |= 0x1;
```

Comment:
we should add a note in the RCUTILS_LOG_SEVERITY that the last bit isn't supposed to be used.
Or maybe use a struct and store both the log level and the flag independently.

Reply:
Let me know if you think we should change to a structure.  I avoided that just for the efficiency, but I can do it if you think that is better.